rcl/test/rcl/test_events.cpp

```diff
@@ -853,3 +854,303 @@ INSTANTIATE_TEST_SUITE_P(
   TestEventFixture,
   ::testing::ValuesIn(get_test_pubsub_incompatible_qos_inputs()),
   TestEventFixture::PrintToStringParamName());
+
+struct EventUserData
+{
+  std::shared_ptr<std::atomic_size_t> event_count;
+};
+
+extern "C"
+{
+void event_callback(const void * user_data, size_t number_of_events)
+{
+  (void)number_of_events;
+  const struct EventUserData * data = static_cast<const struct EventUserData *>(user_data);
+  ASSERT_NE(nullptr, data);
+  (*data->event_count)++;
+}
+}
+
+/*
+ * Basic test of publisher matched/unmatched event
+ */
+TEST_F(TestEventFixture, test_pub_matched_unmatched_event)
+{
+  rcl_ret_t ret;
+
+  // Create one publisher
+  setup_publisher(default_qos_profile);
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  });
+
+  // init publisher event
+  rcl_event_t pub_matched_event = rcl_get_zero_initialized_event();
+  ret = rcl_publisher_event_init(&pub_matched_event, &publisher, RCL_PUBLISHER_MATCHED);
+  ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    ret = rcl_event_fini(&pub_matched_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  });
+  rcl_event_t pub_unmatched_event = rcl_get_zero_initialized_event();
+  ret = rcl_publisher_event_init(&pub_unmatched_event, &publisher, RCL_PUBLISHER_UNMATCHED);
+  ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    ret = rcl_event_fini(&pub_unmatched_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  });
+
+  // init event callback
+  struct EventUserData matched_data;
+  matched_data.event_count = std::make_shared<std::atomic_size_t>(0);
+  struct EventUserData unmatched_data;
+  unmatched_data.event_count = std::make_shared<std::atomic_size_t>(0);
+
+  // rmw_connextdds doesn't support rmw_event_set_callback() interface
+  if (std::string(rmw_get_implementation_identifier()).find("rmw_connextdds") != 0) {
+    ret = rcl_event_set_callback(&pub_matched_event, event_callback, &matched_data);
+    ASSERT_EQ(RMW_RET_OK, ret);
+
+    ret = rcl_event_set_callback(&pub_unmatched_event, event_callback, &unmatched_data);
```

Comment:
I think there should be test case that this is called after creating the subscription. it is expected that callback should not be fired because the matched event has happened before.

Reply:
Yes. I willl add test cases.